src/main/scala/firrtl/options/StageAnnotations.scala

```diff
@@ -14,6 +17,61 @@ sealed trait StageOption { this: Annotation => }
   */
 trait Unserializable { this: Annotation => }
 
+/** Mix-in that lets an [[firrtl.annotations.Annotation Annotation]] serialize itself to a file separate from the output
+  * annotation file.
+  *
+  * This can be used as a mechanism to serialize an [[firrtl.options.Unserializable Unserializable]] annotation or to
+  * write ancillary collateral used by downstream tooling, e.g., a TCL script or an FPGA constraints file. Any
+  * annotations using this mix-in will be serialized by the [[firrtl.options.phases.WriteOutputAnnotations
+  * WriteOutputAnnotations]] phase. This is one of the last phases common to all [[firrtl.options.Stage Stages]] and
+  * should not have to be called/included manually.
+  *
+  * Note: from the perspective of transforms generating annotations that mix-in this trait, the serialized files are not
+  * expected to be available to downstream transforms. Communication of information between transforms must occur
+  * through the annotations that will eventually be serialized to files.
+  */
+trait CustomFileEmission { this: Annotation =>
+
+  /** Output filename where serialized content will be written */
+  protected def baseFileName: String
+
+  /** Optional suffix of the output file */
+  protected def suffix: Option[String]
+
+  /** A method that can convert this annotation to bytes that will be written to a file.
+    *
+    * If you only need to serialize a string, you can use the `getBytes` method:
+    * {{{
+    *  def howToSerialize: Option[Iterable[Byte]] = Some(myString.getBytes)
+    * }}}
+    */
+  def toBytes: Option[Iterable[Byte]]
+
+  /** Optionally, a sequence of annotations that will replace this annotation in the output annotation file.
+    *
+    * A non-None implementation of this method is a mechanism for telling a downstream [[firrtl.options.Stage Stage]]
+    * how to deserialize the information that was serialized to a separate file. For example, if a FIRRTL circuit is
+    * serialized to a separate file, this method could include an input file annotation that a later stage can use to
+    * read the serialized FIRRTL circuit back in.
+    */
+  def replacements(file: File): Option[AnnotationSeq]
```

Comment:
I wonder if we could make this a separate trait, maybe one that extends from the `CustomFileEmission` trait.
Or maybe we could have a default implementation that returns `None`?

Reply:
Default implementation seems good to me. One-off traits to provide partial implementations has seemed attractive before (e.g., `PreservesAll[_]`), but after the headaches with `PreservesAll`, I'm not sure this is actually a good approach.

I'm fixing this up in the rebase. It also occurred to me that `Option[AnnotationSeq]` is weird. I'm trying to do two things here: (1) express the property that there may be no replacements and (2) it's nice to have the replacements in a monad as this lets me concisely `flatMap` over the annotation seq and produce zero or more replacement annotations. `AnnotationSeq` is *almost* a monad (if you call the `toSeq` method on it), so I think it makes sense to just switch the type to `AnnotationSeq`.

Does changing the return type seem reasonable?